argcomplete/__init__.py

```diff
@@ -284,30 +289,46 @@ def collect_completions(self, active_parsers, parsed_args, cword_prefix, debug):
 
         This method is exposed for overriding in subclasses; there is no need to use it directly.
         '''
-        completions = []
-        for parser in active_parsers:
-            debug("Examining parser", parser)
-            for action in parser._actions:
-                debug("Examining action", action)
-                if isinstance(action, argparse._SubParsersAction):
-                    subparser_activated = False
-                    for subparser in action._name_parser_map.values():
-                        if subparser in active_parsers:
-                            subparser_activated = True
-                    if subparser_activated:
-                        # Parent parser completions are not valid in the subparser, so flush them
-                        completions = []
-                    else:
-                        completions += [subcmd for subcmd in action.choices.keys() if subcmd.startswith(cword_prefix)]
-                elif self.always_complete_options or (len(cword_prefix) > 0 and cword_prefix[0] in parser.prefix_chars):
-                    completions += [option for option in action.option_strings if option.startswith(cword_prefix)]
+        def get_subparser_completions(parser, cword_prefix):
+            """ """
+            def filter_aliases(metavar, dest, prefix):
+                if not metavar:
+                    return dest
+
+                a = metavar.replace(',', '').split()
+                return ' '.join(x for x in a if x.startswith(prefix))
+
+            self._display_completions.update(
+                [[filter_aliases(action.metavar, action.dest, cword_prefix),action.help]
+                 for action in parser._get_subactions()])
```

Comment:
The construct that you use to add completions to _display_completions is really hard to read. Please just use a for loop and append to _display_completions. Also, if you're going to use tuples like this, please use NamedTuples so that the semantic meaning of the contents is apparent.